Yatima/Typechecker/Infer.lean

```diff
@@ -20,147 +20,189 @@ In this module the two major functions `check` and `infer` are defined.
     match term with
-    | .lam _ lamName _ _lamDom bod =>
-      match type with
-      | .pi _ _ _ dom img env =>
+    | .lam _ lamName bind lamDom bod =>
+      match type.get with
+      | .pi _ _ dom img env =>
         let lvl := (← read).lvl
-        let var : Thunk Value := {
-          fn := fun _ =>
-            mkVar sorry lamName lvl,
-          repr := "",
-        }
-        let img ← withNewExtendedEnv env var $ eval img
-        -- TODO check that `lamDom` == `dom`
-        -- though this is wasteful, since this would force
-        -- `dom`, which might not need to be evaluated.
-        withExtendedCtx var dom $ check bod img
-      | val => throw $ .notPi (printVal val)
-    | .letE _ _ expType exp bod =>
+        let (lamDom, _) ← isSort lamDom
+        if !(dom.info == lamDom.info) || !(← equal dom.info lvl (← eval lamDom) dom.get) then
+          throw $ .custom "Lambda annotation does not match with its type"
+        let var := mkSusVar (← infoFromType dom) lamName lvl
+        let img := suspend img { ← read with env := env.extendWith var }
+        let bod ← withExtendedCtx var dom $ check bod img
+        pure $ .lam (lamInfo bod.info) lamName bind lamDom bod
+      | val => throw $ .notPi (toString val)
+    | .letE _ name expType exp bod =>
       let (expType, _) ← isSort expType
-      let expType ← eval expType
-      let _ ← check exp expType
-      let exp := suspend exp (← read)
-      withExtendedCtx exp expType $ check bod type
+      let expTypeVal := suspend expType (← read)
+      let exp ← check exp expTypeVal
+      let expVal := suspend exp (← read)
+      let bod ← withExtendedCtx expVal expTypeVal $ check bod type
+      pure $ .letE bod.info name expType exp bod
     | _ =>
       let (term, inferType) ← infer term
-      if !(← equal (← read).lvl type inferType) then
-        throw $ .valueMismatch (printVal inferType) (printVal type)
+      if !(inferType.info == type.info) || !(← equal inferType.info (← read).lvl type.get inferType.get) then
+        throw $ .valueMismatch (toString inferType.get) (toString type.get)
       else
         pure term
 
   /-- Infers the type of `term : Expr`. Returns the expression with flags updated along with the inferred type  -/
-  partial def infer (term : Expr) : TypecheckM (Expr × Value) := do
+  partial def infer (term : Expr) : TypecheckM (Expr × SusValue) := do
     match term with
     | .var _ name idx =>
       let types := (← read).types
       let some type := types.get? idx | throw $ .outOfEnvironmentRange name idx types.length
-      pure (term, type.get)
+      let term := .var (← infoFromType type) name idx
+      pure (term, type)
     | .sort _ lvl =>
-      let lvl := Univ.instBulkReduce (← read).env.univs lvl.succ
-      return (term, Value.sort sorry lvl)
+      let univs := (← read).env.univs
+      let typ := .mk sortInfo ⟨ fun _ => .sort $ .instBulkReduce univs lvl.succ ⟩
+      return (term, typ)
```

Comment:
Do we want to annotate `term` as well? Same question for the `.lit` cases below. Not sure if not doing so would break any invariants.

Reply:
Sorry about the delay, but there's no need to annotate here, since a `Sort lvl` is never a proof, prop, unit or struct